### ISSUE TITLE
[java] AvoidFieldNameMatchingMethodName is case insensitive

### ISSUE THREAD

**Issue by akshayt23** [pmd/pmd#2230]
**Rule:**
AvoidFieldNameMatchingMethodName

**Description:**
The rule seems to be case insensitive and hence, reports a violation for the code sample below. 

**Code Sample demonstrating the issue:**

```
public class Foo implements IdProvider {

    private static final String ID = "foo";

    @Override
    public String id() {
        return ID;
    }

    private void blah() {
        // Another method that uses ID.
    }
}
```

**Running PMD through:** *[Gradle]*

**Comment by adangel** [comment]
I'm unsure, whether we should or should not report a violation here. The rule implementation as now is indeed case-insensitive (https://github.com/pmd/pmd/blob/99fdede0aa7fe2c2bd9f33acb94add409366c979/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidFieldNameMatchingMethodNameRule.java#L48).

In your case, the method name "id()" is problematic, and the rule - if I understand the intention correctly - would suggest to use "getId()" instead - to have a "clear naming". While technically, we would change the implementation to be case sensitive, I don't think, that this would be in the spirit of the rule: The two names would still be very similar...

Would could add a property to enable/disable case-sensitive comparing. Not sure if that's worth, since you can simply suppress this "special case" like this:

```java
@SuppressWarnings("PMD.AvoidFieldNameMatchingMethodName")
private static final String ID = "foo";
```

What do you think?

**Comment by akshayt23** [comment]
@adangel Thanks for the quick response.

Personally, I feel the two names are distinct enough in this case that it shouldn't be a violation. Although yes, technically, they are the same name.

Also, renaming the method is always not under our control; as in this case the `id()` method actually is part of the `IdProvider` interface. So either we'd end up calling the internal constant something else, or suppressing the warning altogether as you pointed out.

**Comment by Fernal73** [comment]
If the class is a generated class, I'd suggest not running PMD on any such classes. That's the practice I employ for Antlr generated classes that generate violations.
If it's not, you can either suppress the violation or rename the field.

**Comment by Fernal73** [comment]
Add another configuration option? More control to the user.

**Comment by Fernal73** [comment]
https://github.com/pmd/pmd/issues/2094

There's a new rule suggested as above. Should this rule be bundled into that?

**Comment by d3bt3ch** [comment]
In case of a builder pattern most of the times the variable names matches the function name and is not a bad code style. So, why is this rule included in PMD?
